sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/WorkerBase.java

```diff
@@ -0,0 +1,186 @@
+
+  protected final BlockingQueue<ReadableSpan> queue;
+
+  protected final AtomicReference<CompletableResultCode> flushRequested = new AtomicReference<>();
+  protected final AtomicBoolean continueWork = new AtomicBoolean(true);
+
+  /**
+   * This a base class useful for all flavors of BatchSpanProcessor.
+   *
+   * @param spanExporter {@link SpanExporter} to use
+   * @param scheduleDelayNanos how much time can pass before Worker sends a batch to exporter
+   * @param maxExportBatchSize max size of a batch
+   * @param exporterTimeoutNanos {@link SpanExporter} timeout
+   * @param queue thread-safe queue for processing spans
+   * @param spanProcessorTypeLabel (ie. queueSize or processedSpans) metric label
+   * @param spanProcessorTypeValue (ie. queueSize or processedSpans) metric value
+   */
+  protected WorkerBase(
+      SpanExporter spanExporter,
+      long scheduleDelayNanos,
+      int maxExportBatchSize,
+      long exporterTimeoutNanos,
+      BlockingQueue<ReadableSpan> queue,
+      String spanProcessorTypeLabel,
+      String spanProcessorTypeValue) {
+    this.spanExporter = spanExporter;
+    this.scheduleDelayNanos = scheduleDelayNanos;
+    this.maxExportBatchSize = maxExportBatchSize;
+    this.exporterTimeoutNanos = exporterTimeoutNanos;
+    this.queue = queue;
+    Meter meter = GlobalMetricsProvider.getMeter("io.opentelemetry.sdk.trace");
+    meter
+        .longValueObserverBuilder("queueSize")
+        .setDescription("The number of spans queued")
+        .setUnit("1")
+        .setUpdater(
+            result ->
+                result.observe(
+                    queue.size(), Labels.of(spanProcessorTypeLabel, spanProcessorTypeValue)))
+        .build();
+    LongCounter processedSpansCounter =
+        meter
+            .longCounterBuilder("processedSpans")
+            .setUnit("1")
+            .setDescription(
+                "The number of spans processed by the BatchSpanProcessor. "
+                    + "[dropped=true if they were dropped due to high throughput]")
+            .build();
+    droppedSpans =
+        processedSpansCounter.bind(
+            Labels.of(spanProcessorTypeLabel, spanProcessorTypeValue, "dropped", "true"));
+    exportedSpans =
+        processedSpansCounter.bind(
+            Labels.of(spanProcessorTypeLabel, spanProcessorTypeValue, "dropped", "false"));
+  }
+
+  /**
+   * Adds a span to worker's internal queue.
+   *
+   * @param span {@link ReadableSpan}
+   */
+  public void addSpan(ReadableSpan span) {
+    if (!queue.offer(span)) {
+      droppedSpans.add(1);
+    }
+  }
+
+  protected abstract Collection<SpanData> getBatch();
+
+  protected void exportCurrentBatch() {
+    Collection<SpanData> batch = getBatch();
+    if (batch.isEmpty()) {
+      return;
+    }
+
+    try {
+      final CompletableResultCode result = spanExporter.export(new ArrayList<>(batch));
+      result.join(exporterTimeoutNanos, TimeUnit.NANOSECONDS);
```

Comment:
Sorry that it took reading code to come up with this basic question - is there a use case for sharing an `ExecutorService` if we're doing blocking I/O? For example, if we try to share a single thread between span processor and interval metric reader, than while we're waiting for spans to be exported, this thread is asleep doing nothing while metrics can't be processed. Or if we use two thread executor, then we may as well have used separate single-thread executors instead of opening up a can of worms for blocking I/O to affect other processes. @Oberon00 what do you think?

We had an asynchronous BSP at one point and for such a processor, sharing an executor makes a lot of sense.

Reply:
> while we're waiting for spans to be exported, this thread is asleep doing nothing while metrics can't be processed

The current implementation of the IntervalMetricReader does metric collection and export on the same thread, so this can really be a problem. But if it did metric export and collection on separate threads, it could share only the metric exporting thread, not the collection thread.

> We had an asynchronous BSP at one point

I think the `join` will need to be removed and the BSP needs to use the `whenComplete` mechanism for this implementation to become more robust (to maintain the timeout, maybe additionally use `schedule` to signal the whenComplete callback to cancel the whenComplete).

This will also mean that we can no longer use `scheduleWithFixedDelay` but need to `schedule()` the next export ourselves whenever the last `whenComplete` is done.

@anuraaga Does that sound reasonable?

Reply:
> it could share only the metric exporting thread, not the collection thread.

I guess even with this, if we're not doing the I/O simultaneously it could cause issues with the export throughput. Pretty sure that's what you're thinking too just confirming :)

> This will also mean that we can no longer use scheduleWithFixedDelay but need to schedule() the next export ourselves whenever the last whenComplete is done.

Yeah this seems ok to me - unless I'm missing something I think it's a prerequisite for having exporters share threads.

Reply:
> it could cause issues with the export throughput

I'm not sure how serious this is. Even if metrics and spans have to wait for each other, it depends on the intervals if this is a problem. In the worst case where the thread is always "busy" (at least waiting for I/O to complete) and metrics and spans cause about the same traffic, even when using two threads (one for metrics, one for spans) you get at most 2x improvement.

Reply:
@anuraaga @Oberon00 thank you for the detailed explanation. I will rework my SpanProcessor.